node/msg_publisher.go

```diff
@@ -0,0 +1,38 @@
+package node
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-filecoin/core"
+	"github.com/filecoin-project/go-filecoin/net/pubsub"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/pkg/errors"
+)
+
+// defaultMessagePublisher publishes messages to a pubsub topic and adds them to a message pool.
+// This is wiring for message publication from the outbox.
+type defaultMessagePublisher struct {
+	network *pubsub.Publisher
+	topic   string
+	pool    *core.MessagePool
+}
+
+func newDefaultMessagePublisher(pubsub *pubsub.Publisher, topic string, pool *core.MessagePool) *defaultMessagePublisher {
+	return &defaultMessagePublisher{pubsub, topic, pool}
+}
+
+func (p *defaultMessagePublisher) Publish(ctx context.Context, message *types.SignedMessage, height uint64) error {
+	encoded, err := message.Marshal()
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal message")
+	}
+
+	if _, err := p.pool.Add(ctx, message); err != nil {
+		return errors.Wrap(err, "failed to add message to message pool")
+	}
+
+	if err = p.network.Publish(p.topic, encoded); err != nil {
+		return errors.Wrap(err, "failed to publish message to network")
+	}
+	return nil
```

Comment:
Looks great, but I'm not totally clear on why this is under `node`. Why not put it with the outbox? It feels like the `node` directory has been sort of devolving into a "misc" folder lately.

Reply:
This is a package-level dependency injection. The outbox doesn't know or care how messages are published to the network. This change removed the import of `net/pubsub` by the `core` package.

It's "wiring" which definitely feels miscellaneous but is what the node package alone should be doing.